### ISSUE TITLE
legacy release 2020.07.xx tracking

### ISSUE THREAD

**Issue by mirabilos** [cgeo/cgeo#8602]
As discussed in https://github.com/cgeo/cgeo/issues/8252 and promised in https://github.com/cgeo/cgeo/issues/8120#issuecomment-643638078 and the following comment, a version for legacy devices (with working OC and Mapsforge v3) should be published that matches the current feature release.

Users will also need instructions of how to upgrade from a previous version (nightly or release) to the new legacy release, i.e. whether a full database and settings dump/export and reload/import is needed, and how to install the new legacy release over F-Droid.

**Comment by Lineflyer** [comment]
The plan is to cut off a legacy release immediately before next time merging `master` to `release` . It will then just contain all what is in the last Play store release.
Once this state has been cut off to the branch `cgeo-legacy` we will do the required tasks as e.g.
- Set appropriate maxSDK
- Implement OC SSL workaround

Mapsforge V3 should be supported anyhow as it is in current Playstore release.
Once published it will be available as normal Google Play update and you will be able to retrieve that via Google Play if minSDK<=Your device API < maxSDK
Upgrade from nightly to that will be not possible (as currently).

Legacy version has not been setup for F-Droid last time we used it, as to that time we did not yet have F-Droid repository in place. But should not be a problem.

**Comment by mirabilos** [comment]
Wasn’t the plan to do this for the 2020.06.xx release already?

I’m kinda stuck on my 2019.12.xx NB…

Anyway, just opened this so that the progress of the legacy release is tracked ANYWHERE at all, and so I get notifications about it ☺

**Comment by Lineflyer** [comment]
> I’m kinda stuck on my 2019.12.xx NB…

Why is that?
The change to higher SDK should have appeared much later. What Android version are you using?

**Comment by moving-bits** [comment]
Change to higher SDK is on master branch only, but IIRC he's stuck to the old version due to SSL handling - this is what the pending fix for legacy tries to mitigate.

**Comment by mirabilos** [comment]
Lars dixit:

>> I’m kinda stuck on my 2019.12.xx NB…
>
>Why is that?

Because at that point I learnt you wanted to drop support for
older Androids and I held off updating.

>The change to higher SDK should have appeared much later.

Yes, it did, but I did not know when exactly.

>What Android version are you using?

4.4

>Change to higher SDK is on master branch only, but IIRC he's stuck to
>the old version due to SSL handling -

That too, but that NB is already past that point, so I’m currently
importing OC via GPX and logging them offline :/

**Comment by Lineflyer** [comment]
Seeing the amount of changes merged and upcoming on `master` maybe we should cut of a beta for a new feature release rather soon, which at the same time triggers to prepare the legacy release.

@moving-bits What do you think? I always tend to keep the amount of changes (and thus possible regressions) low and rather do multiple smaller releases, therefore asking.

**Comment by moving-bits** [comment]
Sounds reasonable to me. If there's nothing urgent with the current release version, we can start cutting off the legacy branch now. I opened a separate issue to discuss which needs to be done for the upcoming feature release before we can prepare a beta.

**Comment by Lineflyer** [comment]
I will do some preparation tonight, i.e.:
- Copy the state of `cgeo-legacy` to another new branch `cgeo-legacy-API9-13` to keep it as a reference for build config
- Force push state of `release` to `cgeo-legacy` to check on CI, whether the legacy jobs as still useable. We can update the branch later on to reflect `release` if needed.

**Comment by Lineflyer** [comment]
The old [legacy job](https://ci.cgeo.org/view/Release_Nightly/job/cgeo%20legacy/) on CI is still working after updateíng post-build actions to the meanwhile modified ftp upload script.

However there are some items not OK or unclear:
- Although I set versionName and versionCode for the build job the final APK uses the current date versionName/versionCode but not the variables defined for the build job. I assume the gradle script does not take the `legacy` path
- For the old cgeo-legacy we used to set minSDK and maxSDK in `AndroidManifest.xml`. This mechanism seems to have changed meanwhile to be use `build.gradle`. How to overcome this and define minSDK and maxSDK accordingly for that branch / that build type

@kumy @mucek4  Help appreciated, as my knowledge of the CI is limited.

**Comment by Lineflyer** [comment]
@kumy Do you remember the logic here (in case you were involved last time?)?

**Comment by kumy** [comment]
Sorry I missed your last mention.

@Lineflyer from what I understand, the gradle target in the old [legacy job](https://ci.cgeo.org/view/Release_Nightly/job/cgeo%20legacy/) was wrong. ie `assembleBasicRelease` instead of `assembleBasicLegacy`.

The source code related to this:
https://github.com/cgeo/cgeo/blob/93350adcaac9d44924513d154bf89de10b429aba/main/build.gradle#L224-L233

I've made the changes and launched another [build #24](https://ci.cgeo.org/view/Release_Nightly/job/cgeo%20legacy/24/) (upload to ftp has been temporary disabled - feel free to unlock upload if we're ok)
Changes:
![Screenshot from 2020-07-17 10-23-00](https://user-images.githubusercontent.com/176794/87764974-8175b780-c817-11ea-92e1-3849239e718a.png)

Apk data seems now ok:
https://ci.cgeo.org/view/Release_Nightly/job/cgeo%20legacy/ws/main/build/outputs/apk/basic/legacy/output.json
![Screenshot from 2020-07-17 10-18-09](https://user-images.githubusercontent.com/176794/87764537-dc5adf00-c816-11ea-93d4-d125df0f7e3f.png)

The built apk is temporarily available at (temporarily as it will be overwritten on next build) https://ci.cgeo.org/view/Release_Nightly/job/cgeo%20legacy/ws/main/build/outputs/apk/basic/legacy/cgeo-legacy.apk

@Lineflyer may you please cross check :)

I have also adapted the ftp upload part to match paths and filenames, but not tested it.

**Comment by Lineflyer** [comment]
@kumy 
Thanks. Just tested it and works as expected. I also reenabled the ftp upload which also works well.
Thanks for your detailed report, which made me learn a lot for future.

**Comment by Lineflyer** [comment]
According to https://developer.android.com/guide/topics/manifest/uses-sdk-element using `maxSDKversion` is not mandatory but optional meanwhile.
As far as I understood the alternative logic regarding Google Play [here](https://developer.android.com/google/play/publishing/multiple-apks) we just need to make sure, that an APK with higher `minSDKversion` also has a higher version code, than the APK for a lower `minSDKversion`.
While this sounds logical and feasible, I wonder what needs to be done in a scenario, when we want provide an update to the legacy version only, as I do which versioncode to use then.
Is it: versionCode_oldlegacy < versionCode_newlegacy < versionCode_normalrelease

**Comment by moving-bits** [comment]
> Is it: versionCode_oldlegacy < versionCode_newlegacy < versionCode_normalrelease

After having read above articles I would agree.

**Comment by bekuno** [comment]
As discussed in https://github.com/cgeo/WhereYouGo/pull/183 - the `maxSDKversion` is optional, but we can/should set it to stay independent from release branch.

**Comment by Lineflyer** [comment]
> As discussed in cgeo/WhereYouGo#183 - the maxSDKversion is optional, but we can/should set it to stay independent from release branch.

Will do that.

**Comment by mirabilos** [comment]
> As discussed in cgeo/WhereYouGo#183 - the maxSDKversion is optional,
> but we can/should set it to stay independent from release branch.

Doesn’t make this it difficult to test/use the legacy version
on a device with modern Android?

**Comment by Lineflyer** [comment]
No, AFAIK this is only evaluated by GooglePlay to distribute matching versions to different API versions. You can manually install any APK you want.

**Comment by Lineflyer** [comment]
A release candidate for a new legacy version has been published as beta version on Google Play: 2020.07.24-RC-legacy
Android 4.1 - 4.4 devices from users on beta channel should receive this update normally now.

@mirabilos 
Feel free to try it out via Google Play (might need some hours until available).
There is no fix for https issues yet included, just a proof of concept for the legacy branch as such.

@jonas-koeritz 
Might you want to apply your proposal in #8139 targeted to branch `cgeo-legacy`?

**Comment by mirabilos** [comment]
Hi Lars,

>@mirabilos Feel free to try it out via Google Play (might need some
>@hours until available).

thanks, nice, but, as I said, I cannnot run Google Play on my device
as it makes it too slow.

>There is no fix for https issues yet included, just a proof of concept
>for the legacy branch as such.

OK, works for me; as such, it’s not a regression against the Dec 2019 NB
I currently use.

Do I need to make a DB and settings dump, uninstall the old application,
install the new one, and import the dump to switch? Or are they using
different application names so I can install them in parallel (and if so,
do they use different locations for the DB and settings)?

**Comment by Lineflyer** [comment]
For information:
The first legacy release has been promoted from beta state on normal Google Play release and should be available soon.
Version name is 2020.08.04-legacy and it will be available for API 16-20 (Android 4.1 to 4.4).

Fix for https (see issue #7497) is not yet included.

**Comment by Lineflyer** [comment]
@mirabilos Download is also possible outside of Google Play here:
https://github.com/cgeo/cgeo/releases/tag/market_20200726_legacy

**Comment by mirabilos** [comment]
Lars dixit:

>@mirabilos Download is also possible outside of Google Play here:
>https://github.com/cgeo/cgeo/releases/tag/market_20200726_legacy

OK.

My other questions (regarding installation/upgrade and data migration
method) was not answered yet though.

**Comment by Lineflyer** [comment]
> 
Do I need to make a DB and settings dump, uninstall the old application,
install the new one, and import the dump to switch? Or are they using
different application names so I can install them in parallel (and if so,
do they use different locations for the DB and settings)?

It depends:
- If you currently use a nightly, you have to deinstall and install the legacy version (no upgrade path)
- If you happen to use a nightly of 2019 you can probably restore the backup (which you need to perform prior to deinstallation of the nightly). 
- All versions (normal release, nightly, legacy) share the same application ID / package name, thus no parallel installations possible

**Comment by mirabilos** [comment]
Lars dixit:

>>Do I need to make a DB and settings dump, uninstall the old
>>application, install the new one, and import the dump to switch? Or
>>are they using different application names so I can install them in
>>parallel (and if so, do they use different locations for the DB and
>>settings)?
>
>It depends:
>- If you currently use a nightly, you have to deinstall and install the
>  legacy version (no upgrade path)
>- If you happen to use a nightly of 2019 you can probably restore the
>  backup (which you need to perform prior to deinstallation of the
>  nightly).
>- All versions (normal release, nightly, legacy) share the same
>  application ID / package name, thus no parallel installations
>  possible

OK, thanks, this is exactly what I was needing, I’ll try this then.

bye,
//mirabilos
-- 
<diogenese> Beware of ritual lest you forget the meaning behind it.
<igli> yeah but it means if you really care about something, don't
    ritualise it, or you will lose it. don't fetishise it, don't
    obsess. or you'll forget why you love it in the first place.
